tpl/template.go

```diff
@@ -447,33 +447,43 @@ func (t *GoHTMLTemplate) loadTemplates(absPath string, prefix string) {
 				}
 				if needsBase {
 
+					layoutDir := helpers.GetLayoutDirPath()
+					currBaseFilename := fmt.Sprintf("%s-%s", helpers.Filename(path), baseFileName)
+					templateDir := filepath.Dir(path)
+					themeDir := filepath.Join(helpers.GetThemeDir())
+					relativeThemeLayoutsDir := filepath.Join(helpers.GetRelativeThemeDir(), "layouts")
+
+					var baseTemplatedDir string
+
+					if strings.HasPrefix(templateDir, relativeThemeLayoutsDir) {
+						baseTemplatedDir = strings.TrimPrefix(templateDir, relativeThemeLayoutsDir)
+					} else {
+						baseTemplatedDir = strings.TrimPrefix(templateDir, layoutDir)
+					}
+
+					baseTemplatedDir = strings.TrimPrefix(baseTemplatedDir, helpers.FilePathSeparator)
+
+					var pathsToCheck []string
+
 					// Look for base template in the follwing order:
 					//   1. <current-path>/<template-name>-baseof.<suffix>, e.g. list-baseof.<suffix>.
 					//   2. <current-path>/baseof.<suffix>
 					//   3. _default/<template-name>-baseof.<suffix>, e.g. list-baseof.<suffix>.
 					//   4. _default/baseof.<suffix>
-					//   5. <themedir>/layouts/_default/<template-name>-baseof.<suffix>
-					//   6. <themedir>/layouts/_default/baseof.<suffix>
-
-					currBaseFilename := fmt.Sprintf("%s-%s", helpers.Filename(path), baseFileName)
-					templateDir := filepath.Dir(path)
-					themeDir := helpers.GetThemeDir()
-
-					pathsToCheck := []string{
-						filepath.Join(templateDir, currBaseFilename),
-						filepath.Join(templateDir, baseFileName),
-						filepath.Join(absPath, "_default", currBaseFilename),
-						filepath.Join(absPath, "_default", baseFileName),
-						filepath.Join(themeDir, "layouts", "_default", currBaseFilename),
-						filepath.Join(themeDir, "layouts", "_default", baseFileName),
-					}
+					// For each of the steps above, it will first look in the project, then, if theme is set,
+					// in the theme's layouts folder.
+					pathsToCheck = appendBasePathsToCheck(pathsToCheck, layoutDir, themeDir, []string{baseTemplatedDir, currBaseFilename})
```

Comment:
Could this be refactored to avoid the path building until it's necessary (avoiding allocs & cycles if the template is found early)?

```go
    pairsToCheck := [][]string{
        []string{baseTemplatedDir, currBaseFilename},
        []string{baseTemplatedDir, baseFileName},
        []string{"_default", currBaseFilename},
        []string{"_default", baseFileName},
    }

OuterLoop:
    for _, pair := range pairsToCheck {
        pathsToCheck := basePathsToCheck(pair, layoutDir, themeDir)
        for _, pathToCheck := range pathsToCheck {
            if ok, err := helpers.Exists(pathToCheck, hugofs.Source()); err == nil && ok {
                baseTemplatePath = pathToCheck
                break OuterLoop
            }
        }
    }

// refactor appendBasePathsToCheck as basePathsToCheck
```